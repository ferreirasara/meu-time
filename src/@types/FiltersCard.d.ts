export type FiltersCardProps = {
  selectedCountry: string
  handleSelectCountry: (country: string) => void
  selectedLeague: number
  handleSelectLeague: (league: number) => void
  selectedSeason: number
  handleSelectSeason: (season: number) => void
  selectedTeam: number
  handleSelectTeam: (team: number) => void
}