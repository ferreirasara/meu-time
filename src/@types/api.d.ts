export type Team = {
  team: {
    id: number,
    name: string,
    code: string,
    country: string,
    founded: number,
    national: boolean,
    logo: string,
  },
  venue: {
    id: number,
    name: string,
    address: string,
    city: string,
    capacity: number,
    surface: string,
    image: string,
  },
}

export type Country = {
  name: string,
  code?: string,
  flag?: string,
}

export type League = {
  country: Country,
  league: {
    id: number,
    name: string,
    type: string,
    logo: string,
  }
  seasons: {
    year: number,
    start: string,
    end: string,
    current: boolean,
    coverage: {
      fixtures: {
        events: boolean,
        lineups: boolean,
        statistics_fixtures: boolean,
        statistics_players: boolean
      },
      standings: boolean,
      players: boolean,
      top_scorers: boolean,
      top_assists: boolean,
      top_cards: boolean,
      injuries: boolean,
      predictions: boolean,
      odds: boolean
    }
  }[],
}