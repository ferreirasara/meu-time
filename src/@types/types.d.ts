export type BaseData = {
  lastUpdated?: Date,
  countries?: Country[],
  leagues?: League[],
  seasons?: number[],
}

export type UserData = {
  email: string,
  firstname: string,
  lastname: string,
}

export type RequestData = {
  current: number,
  limitDay: number,
}

export type PlayersTableData = {
  id: number,
  name: string,
  age: number,
  nationality: string,
  photo: string,
}

export type PlayersTableProps = {
  selectedLeague: number
  selectedSeason: number
  selectedTeam: number
  handleError: (message: string) => void
}

export type GoalsChartProps = {
  data?: GeneralStatsData,
  loading: boolean,
}

export type GeneralStatsData = {
  totalDraws: number,
  totalPlayed: number,
  totalWins: number,
  totalLoses: number,
  mostUsedFormation: {
    formation: string,
    played: number,
  },
  goals: {
    against: {
      minute: {
        "0-15": { total: number, percentage: string },
        "16-30": { total: number, percentage: string },
        "31-45": { total: number, percentage: string },
        "46-60": { total: number, percentage: string },
        "61-75": { total: number, percentage: string },
        "76-90": { total: number, percentage: string },
        "91-105": { total: number, percentage: string },
        "106-120": { total: number, percentage: string }
      },
    },
    for: {
      minute: {
        "0-15": { total: number, percentage: string },
        "16-30": { total: number, percentage: string },
        "31-45": { total: number, percentage: string },
        "46-60": { total: number, percentage: string },
        "61-75": { total: number, percentage: string },
        "76-90": { total: number, percentage: string },
        "91-105": { total: number, percentage: string },
        "106-120": { total: number, percentage: string }
      },
    },
  }
}

export type GeneralStatsProps = {
  selectedLeague: number
  selectedSeason: number
  selectedTeam: number
  handleError: (message: string) => void
}

export type FiltersCardProps = {
  selectedCountry: string
  handleSelectCountry: (country: string) => void
  selectedLeague: number
  handleSelectLeague: (league: number) => void
  selectedSeason: number
  handleSelectSeason: (season: number) => void
  selectedTeam: number
  handleSelectTeam: (team: number) => void
  localStorageDataLoading: boolean
  handleError: (message: string) => void
}

export type Team = {
  team: {
    id: number,
    name: string,
    code: string,
    country: string,
    founded: number,
    national: boolean,
    logo: string,
  },
  venue: {
    id: number,
    name: string,
    address: string,
    city: string,
    capacity: number,
    surface: string,
    image: string,
  },
}

export type Country = {
  name: string,
  code?: string,
  flag?: string,
}

export type League = {
  country: Country,
  league: {
    id: number,
    name: string,
    type: string,
    logo: string,
  }
  seasons: {
    year: number,
    start: string,
    end: string,
    current: boolean,
    coverage: {
      fixtures: {
        events: boolean,
        lineups: boolean,
        statistics_fixtures: boolean,
        statistics_players: boolean
      },
      standings: boolean,
      players: boolean,
      top_scorers: boolean,
      top_assists: boolean,
      top_cards: boolean,
      injuries: boolean,
      predictions: boolean,
      odds: boolean
    }
  }[],
}

export type Player = {
  player: {
    id: number,
    name: string,
    firstname: string,
    lastname: string,
    age: number,
    birth: {
      date: string,
      place: string,
      country: string,
    },
    nationality: string,
    height: string,
    weight: string,
    injured: boolean,
    photo: string,
  },
  // statistics: {
  //   "team": {
  //     "id": 147,
  //     "name": "Coritiba",
  //     "logo": "https://media-1.api-sports.io/football/teams/147.png"
  //   },
  //   "league": {
  //     "id": 71,
  //     "name": "Serie A",
  //     "country": "Brazil",
  //     "logo": "https://media-2.api-sports.io/football/leagues/71.png",
  //     "flag": "https://media-3.api-sports.io/flags/br.svg",
  //     "season": 2016
  //   },
  //   "games": {
  //     "appearences": 6,
  //     "lineups": 6,
  //     "minutes": 438,
  //     "number": null,
  //     "position": "Defender",
  //     "rating": null,
  //     "captain": false
  //   },
  //   "substitutes": {
  //     "in": 0,
  //     "out": 3,
  //     "bench": 0
  //   },
  //   "shots": {
  //     "total": null,
  //     "on": null
  //   },
  //   "goals": {
  //     "total": 0,
  //     "conceded": null,
  //     "assists": null,
  //     "saves": null
  //   },
  //   "passes": {
  //     "total": null,
  //     "key": null,
  //     "accuracy": null
  //   },
  //   "tackles": {
  //     "total": null,
  //     "blocks": null,
  //     "interceptions": null
  //   },
  //   "duels": {
  //     "total": null,
  //     "won": null
  //   },
  //   "dribbles": {
  //     "attempts": null,
  //     "success": null,
  //     "past": null
  //   },
  //   "fouls": {
  //     "drawn": null,
  //     "committed": null
  //   },
  //   "cards": {
  //     "yellow": 0,
  //     "yellowred": 0,
  //     "red": 0
  //   },
  //   "penalty": {
  //     "won": null,
  //     "commited": null,
  //     "scored": null,
  //     "missed": null,
  //     "saved": null
  //   }
  // }[]
}

export type TeamStats = {
  lineups: {
    formation: string,
    played: number
  }[],
  fixtures: {
    draws: { home: number, away: number, total: number }
    loses: { home: number, away: number, total: number },
    played: { home: number, away: number, total: number },
    wins: { home: number, away: number, total: number },
  },
  goals: {
    against: {
      minute: {
        "0-15": { total: number, percentage: string },
        "16-30": { total: number, percentage: string },
        "31-45": { total: number, percentage: string },
        "46-60": { total: number, percentage: string },
        "61-75": { total: number, percentage: string },
        "76-90": { total: number, percentage: string },
        "91-105": { total: number, percentage: string },
        "106-120": { total: number, percentage: string }
      },
    },
    for: {
      minute: {
        "0-15": { total: number, percentage: string },
        "16-30": { total: number, percentage: string },
        "31-45": { total: number, percentage: string },
        "46-60": { total: number, percentage: string },
        "61-75": { total: number, percentage: string },
        "76-90": { total: number, percentage: string },
        "91-105": { total: number, percentage: string },
        "106-120": { total: number, percentage: string }
      },
    },
  }
}

export type InfoModalProps = {
  infoModalOpen: boolean
  onClose: () => void
}