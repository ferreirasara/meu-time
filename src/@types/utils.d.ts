export type BaseData = {
  lastUpdated?: Date,
  countries?: Country[],
  leagues?: League[],
  seasons?: number[],
}