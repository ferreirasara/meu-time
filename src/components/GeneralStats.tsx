import { Col, Divider, List, Row, Typography } from "antd"
import { useEffect, useState } from "react";
import { getGeneralStatsFromAPI } from "../utils/utils";
import { GoalsChart } from "./GoalsChart";
import { GeneralStatsData, GeneralStatsProps } from "../@types/types";

export const GeneralStats = ({
  selectedLeague,
  selectedSeason,
  selectedTeam,
  handleError,
}: GeneralStatsProps) => {
  const [loading, setLoading] = useState<boolean>(false);
  const [data, setData] = useState<GeneralStatsData>();

  useEffect(() => {
    if (selectedLeague && selectedSeason && selectedTeam) {
      setLoading(true);
      getGeneralStatsFromAPI(selectedLeague, selectedSeason, selectedTeam)
        .then(result => setData(({
          goals: result.goals,
          mostUsedFormation: {
            formation: result?.lineups?.[0]?.formation,
            played: result?.lineups?.[0]?.played,
          },
          totalDraws: result?.fixtures?.draws?.total,
          totalLoses: result?.fixtures?.loses?.total,
          totalPlayed: result?.fixtures?.played?.total,
          totalWins: result?.fixtures?.wins?.total,
        })))
        .catch(e => handleError(e?.toString()))
      setLoading(false);
    }
  }, [selectedLeague, selectedSeason, selectedTeam, handleError])

  return <Row gutter={[16, 16]}>
    <Col span={12}>
      <div>
        <Divider>Estatísticas Gerais</Divider>
        <List bordered loading={loading}>
          <List.Item>
            <Typography.Text strong>Formação mais utilizada</Typography.Text>: {data?.mostUsedFormation?.formation} (usada {data?.mostUsedFormation?.played} vezes)
          </List.Item>
          <List.Item>
            <Typography.Text strong>Total de jogos</Typography.Text>: {data?.totalPlayed}
          </List.Item>
          <List.Item>
            <Typography.Text strong>Total de vitórias</Typography.Text>: {data?.totalWins}
          </List.Item>
          <List.Item>
            <Typography.Text strong>Total de derrotas</Typography.Text>: {data?.totalLoses}
          </List.Item>
          <List.Item>
            <Typography.Text strong>Total de empates</Typography.Text>: {data?.totalDraws}
          </List.Item>
        </List>
      </div>
    </Col>
    <Col span={12}>
      <GoalsChart
        data={data}
        loading={loading}
      />
    </Col>
  </Row>
}