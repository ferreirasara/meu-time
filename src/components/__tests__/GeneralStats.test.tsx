import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import { GeneralStats } from '../GeneralStats';
import { GeneralStatsProps } from '../../@types/types';

const props: GeneralStatsProps = {
  selectedLeague: 0,
  selectedSeason: 0,
  selectedTeam: 0,
  handleError: jest.fn(),
}

describe('GeneralStats', () => {
  beforeAll(() => {
    window.matchMedia = (query) => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // deprecated
      removeListener: jest.fn(), // deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })
  })

  it('match snapshot', () => {
    const generaLStats = renderer.create(<GeneralStats {...props} />).toJSON();
    expect(generaLStats).toMatchSnapshot();
  });

  it('renders stats list', () => {
    const { getByText } = render(<GeneralStats {...props} />);

    const mostUsedFormationEl = getByText(/Formação mais utilizada/);
    expect(mostUsedFormationEl).toBeInTheDocument();

    const totalPlayedEl = getByText(/Total de jogos/);
    expect(totalPlayedEl).toBeInTheDocument();

    const totalWinsEl = getByText(/Total de vitórias/);
    expect(totalWinsEl).toBeInTheDocument();

    const totalLosesEl = getByText(/Total de derrotas/);
    expect(totalLosesEl).toBeInTheDocument();

    const totalDraws = getByText(/Total de empates/);
    expect(totalDraws).toBeInTheDocument();
  });

  it('renders canvas chart', () => {
    const { getByTestId } = render(<GeneralStats {...props} />);

    const chartEl = getByTestId('goals-chart');
    expect(chartEl).toBeInTheDocument();
  });
})