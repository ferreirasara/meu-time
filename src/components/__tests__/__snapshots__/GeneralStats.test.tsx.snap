// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeneralStats match snapshot 1`] = `
<div
  className="ant-row css-dev-only-do-not-override-15rg2km"
  style={
    Object {
      "marginBottom": -8,
      "marginLeft": -8,
      "marginRight": -8,
      "marginTop": -8,
    }
  }
>
  <div
    className="ant-col ant-col-12 css-dev-only-do-not-override-15rg2km"
    style={
      Object {
        "paddingBottom": 8,
        "paddingLeft": 8,
        "paddingRight": 8,
        "paddingTop": 8,
      }
    }
  >
    <div>
      <div
        className="ant-divider css-dev-only-do-not-override-15rg2km ant-divider-horizontal ant-divider-with-text ant-divider-with-text-center"
        role="separator"
      >
        <span
          className="ant-divider-inner-text"
          style={Object {}}
        >
          Estatísticas Gerais
        </span>
      </div>
      <div
        className="ant-list ant-list-split ant-list-bordered css-dev-only-do-not-override-15rg2km"
      >
        <div
          className="ant-spin-nested-loading css-dev-only-do-not-override-15rg2km"
        >
          <div
            className="ant-spin-container"
          >
            <li
              className="ant-list-item ant-list-item-no-flex"
            >
              <span
                className="ant-typography css-dev-only-do-not-override-15rg2km"
                style={
                  Object {
                    "WebkitLineClamp": undefined,
                  }
                }
              >
                <strong>
                  Formação mais utilizada
                </strong>
              </span>
              : 
               (usada 
               vezes)
            </li>
            <li
              className="ant-list-item ant-list-item-no-flex"
            >
              <span
                className="ant-typography css-dev-only-do-not-override-15rg2km"
                style={
                  Object {
                    "WebkitLineClamp": undefined,
                  }
                }
              >
                <strong>
                  Total de jogos
                </strong>
              </span>
              : 
            </li>
            <li
              className="ant-list-item ant-list-item-no-flex"
            >
              <span
                className="ant-typography css-dev-only-do-not-override-15rg2km"
                style={
                  Object {
                    "WebkitLineClamp": undefined,
                  }
                }
              >
                <strong>
                  Total de vitórias
                </strong>
              </span>
              : 
            </li>
            <li
              className="ant-list-item ant-list-item-no-flex"
            >
              <span
                className="ant-typography css-dev-only-do-not-override-15rg2km"
                style={
                  Object {
                    "WebkitLineClamp": undefined,
                  }
                }
              >
                <strong>
                  Total de derrotas
                </strong>
              </span>
              : 
            </li>
            <li
              className="ant-list-item ant-list-item-no-flex"
            >
              <span
                className="ant-typography css-dev-only-do-not-override-15rg2km"
                style={
                  Object {
                    "WebkitLineClamp": undefined,
                  }
                }
              >
                <strong>
                  Total de empates
                </strong>
              </span>
              : 
            </li>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="ant-col ant-col-12 css-dev-only-do-not-override-15rg2km"
    style={
      Object {
        "paddingBottom": 8,
        "paddingLeft": 8,
        "paddingRight": 8,
        "paddingTop": 8,
      }
    }
  >
    <div>
      <div
        className="ant-divider css-dev-only-do-not-override-15rg2km ant-divider-horizontal ant-divider-with-text ant-divider-with-text-center"
        role="separator"
      >
        <span
          className="ant-divider-inner-text"
          style={Object {}}
        >
          Gols por minuto de jogo
        </span>
      </div>
      <canvas
        data-testid="goals-chart"
        height={150}
        role="img"
        width={300}
      />
    </div>
  </div>
</div>
`;
